### PR TITLE
Implemented dshiftl

### PR DESCRIPTION
**Pull request by parth121101**
Fixes: #3572

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -1062,6 +1063,52 @@ namespace Shiftl {
...
+        int64_t val2 = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[2])->m_n;
+        int kind1 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
+        int kind2 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_type);
+        if(kind1 != kind2) {
+            append_error(diag, "'j' argument of 'dshiftl' intrinsic at (1) must be the same type and kind as 'i'", loc);
+            return nullptr;
+        }
+        if(shift < 0){
+            append_error(diag, "The shift argument of 'dshiftl' intrinsic at (1) must be non-negative", loc);
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:1080]: Let's improve the error messages:

* Do not use `(1)`, since we don't show that in the error display
* In ASR, we do not have names of arguments, so just say "first" and "second" argument

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -1062,6 +1063,52 @@ namespace Shiftl {
...
+                ASR::ttype_t* t1, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
+        int64_t val1 = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t val2 = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[2])->m_n;
+        int kind1 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
+        int kind2 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_type);
+        if(kind1 != kind2) {
+            append_error(diag, "'j' argument of 'dshiftl' intrinsic at (1) must be the same type and kind as 'i'", loc);
+            return nullptr;
+        }
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:1078]: We might not have a mechanism for this yet, but ideally this check should be handled in ASR verify(). In evaluation we should assume the ASR is valid, which in this case means `kind1 == kind2`. We can do this later.

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -1062,6 +1063,52 @@ namespace Shiftl {
...
+        /*
+        * r = Dshiftl(x, y, shift)
+        * r = x << shift | y >> (32 - shift)
+        */
+        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[2], ASR::cast_kindType::IntegerToInteger, return_type, nullptr));
+        body.push_back(al, b.Assignment(result, i_BitLshift(args[0], cast, return_type)));
+        body.push_back(al, b.Assignment(result, i_BitOr(result, i_BitRshift(args[1], i_tSub(i(32, return_type), args[2], return_type), return_type), return_type)));
+        ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
+        scope->add_symbol(fn_name, f_sym);
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:1105]: Put an ASR `If` statement to check if`shift` is negative, and if so, add ErrorStop node here.

Comment (parth121101) [reply src/libasr/pass/intrinsic_functions.h:1106]:
```
cat a.f90
program CheckShift
    integer::x,y,s 
    x = 7
    y = 12
    s = -2
    print *, dshiftl(x,y,s)
end program CheckShift

gfortran a.f90 && ./a.out
 -1073741821

lfortran a.f90
-1073741821
```

@certik gfortran does not throw an error in case of runtime variables. So I think we don't need to give an error message for runtime implementation.

File: src/libasr/pass/intrinsic_functions.h [hunk 4]
```diff
@@ -1063,6 +1064,52 @@ namespace Shiftl {
...
+
+        static ASR::expr_t *eval_Dshiftl(Allocator &al, const Location &loc,
+                ASR::ttype_t* t1, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
+        int64_t val1 = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t val2 = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[2])->m_n;
+        int kind1 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
+        int kind2 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_type);
+        if(kind1 != kind2) {
+            append_error(diag, "The kind of first argument of 'dshiftl' intrinsic must be the same as second arguement", loc);
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:1077]:
```suggestion
            append_error(diag, "The kind of first argument of 'dshiftl' intrinsic must be the same as second argument", loc);
```

File: integration_tests/intrinsics_171.f90 [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    print *, kind(dshiftl(10,12,7))
+    if(kind(dshiftl(10,12,7)) /= 4) error stop
+
+    print *, kind(dshiftl(i,j,shift1))
+    if(kind(dshiftl(i,j,shift1)) /= 4) error stop
+
+    print *, kind(dshiftl(a,b,shift2))
+    if(kind(dshiftl(a,b,shift2)) /= 8) error stop
+
+end program intrinsics_171
```
Comment (HarshitaKalani) [inline integration_tests/intrinsics_171.f90:41]: Can you add negative values for `i` and `j`? 
Also,I think we can add an error test to check if the code throws error for negative values of `shift`.

File: src/libasr/pass/intrinsic_functions.h [hunk 6]
```diff
@@ -1063,6 +1064,52 @@ namespace Shiftl {
...
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /*
+        * r = Dshiftl(x, y, shift)
+        * r = x << shift | y >> (32 - shift)
+        */
+        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[2], ASR::cast_kindType::IntegerToInteger, return_type, nullptr));
+        body.push_back(al, b.Assignment(result, i_BitLshift(args[0], cast, return_type)));
+        body.push_back(al, b.Assignment(result, i_BitOr(result, i_BitRshift(args[1], i_tSub(i(32, return_type), args[2], return_type), return_type), return_type)));
+        ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:1105]: Also this implementation only tackles 32 bit integers.

**Comment by certik** [comment]
I think this was implemented in https://github.com/lfortran/lfortran/pull/3602.
